---
title: "powsimR simulations"
output: github_document
---

```{r setup, eval = TRUE, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      root.dir = "/home/vieth/Gits/scRNA-seq-pipelines/simulation/")
library(powsimRDev)

```

# Estimating parameters

The first step is the estimation of parameters needed for simulations, both for genes and spike-ins. We will use a random subsample of 25 cells and 5000 genes from SCRB-seq data processed with zUMIs using STAR aligner with gencode annotation.

```{r}
InputData <- readRDS(file = "SCRBseq.rds")
estparam.gene <- estimateParam(countData = InputData$GeneUMIs, 
                          readData = InputData$GeneReads, 
                          Lengths = InputData$GeneLengths,
                          RNAseq = "singlecell", Protocol = "UMI", Distribution = "NB", 
                          Normalisation = "scran", 
                          GeneFilter = 0.1, SampleFilter = 3)

```

```{r, message=F, error=F,tidy=T}
estparam.spike <- estimateSpike(spikeData = InputData$SpikeCounts, 
                                spikeInfo = InputData$SpikeInfo, 
                                Normalisation = 'depth')
```


```{r, fig.width = 15, fig.height = 15}
plotParam(estparam.gene)
```


# Set up simulations

We will define a small example setup defining differential expression patterns and sample size setup.

```{r}
lfc.gamma.narrow.asym = function(x) sample(c(-1,1), size=x, prob = c(0.25,0.75),replace=T)*rgamma(x, shape = 1, rate = 2)
setup <- Setup(ngenes = 1000, nsims = 10, 
               p.DE = 0.1, pLFC = lfc.gamma.narrow.asym,
               n1 = c(96, 384), n2 = c(96, 384), 
               LibSize = 'equal', DropGenes = TRUE, 
               estParamRes = estparam.gene, estSpikeRes = estparam.spike)
```

# Run simulations

We run the simulations using our recommended pipeline.

```{r}
simres <- simulateDE(SetupRes = setup,
                     Normalisation = "scran", Label = "clustering", 
                     DEmethod = "limma-trend")
```

# Differential Expression Evaluation

```{r}
eval.de <- evaluateDE(simRes = simres, alpha.type = "adjusted", MTC = "BY")
```

```{r, fig.height=10}
plotEvalDE(eval.de, rate = "marginal")
```

